### PR TITLE
feat(node-runtime-worker-thread): Allow to interrupt evalutate calls COMPASS-4562

### PR DESCRIPTION
**Pull request by gribnoysup**
Adds interruption support for both async and sync evaluations for the runtime:

Kapture 2021-02-12 at 17 52 14

~I wanted to include Ctrl-C support for browser repl in the same PR, but ace editor turns out to be a pain to re-configure and I want to open a PR for review before I will be off for NHTT next week, so this is part one that just adds support to the runtime, I'll add UI to be able to use the interruption in a separate PR~ Undocumented properties FTW

### CODE REVIEW

File: packages/node-runtime-worker-thread/src/lock.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export class Lock {
+  private static UNLOCK_TOKEN: UNLOCKED = 'UNLOCKED';
+
+  private promise: Promise<UNLOCKED> | null = null;
+
+  private resolve: ((...args: any[]) => any) | null = null;
+
+  lock(): Promise<UNLOCKED> {
+    this.promise = new Promise((resolve) => {
```
Comment (addaleax) [inline packages/node-runtime-worker-thread/src/lock.ts:11]: Should this make sure that it’s unlocked before?

**Review by addaleax — APPROVED** [review]

File: packages/node-runtime-worker-thread/src/lock.ts (hunk 1 above)
Comment (gribnoysup) [reply packages/node-runtime-worker-thread/src/lock.ts:11]: Yeah, good point, otherwise it can get locked forever in a confusing way, will change

Comment (gribnoysup) [reply packages/node-runtime-worker-thread/src/lock.ts:15]: Fixed in 6680e42

**Comment by gribnoysup** [comment]
@addaleax so sorry, I added some changes after the review as I wasn't anticipating actually implementing Ctrl-C today and opened a PR but then managed to do it (pinging in case you want to give it another look because of that)

**Review by addaleax — APPROVED** [review]
Nice! And don’t worry, it’s actually pretty cool to have it this soon :)

**Comment by addaleax** [comment]
@gribnoysup I guess this is ready to merge? :)

**Comment by gribnoysup** [comment]
Yup, merging! Thanks for the ping

**Pull request merged**